stylegan_code_finder/networks/trans_u_net/utils.py

```diff
@@ -2,22 +2,39 @@
 import torch.nn as nn
 
 
-# This file is taken from https://github.com/Beckschen/TransUNet
-
-class DiceLoss(nn.Module):
-    def __init__(self, n_classes):
-        super(DiceLoss, self).__init__()
+class LossFunction(nn.Module):
+    def __init__(self, n_classes: int):
+        super().__init__()
         self.n_classes = n_classes
 
-    def _one_hot_encoder(self, input_tensor):
+    def _one_hot_encoder(self, input_tensor: torch.Tensor):
         tensor_list = []
         for i in range(self.n_classes):
             temp_prob = input_tensor == i  # * torch.ones_like(input_tensor)
             tensor_list.append(temp_prob.unsqueeze(1))
         output_tensor = torch.cat(tensor_list, dim=1)
         return output_tensor.float()
 
-    def _dice_loss(self, score, target):
+    def _metric_calculation(self, score, target):
+        raise NotImplementedError
+
+    def forward(self, inputs: torch.Tensor, target: torch.Tensor, weight=None, softmax=False) -> float:
+        if softmax:
+            inputs = torch.softmax(inputs, dim=1)
+        target = self._one_hot_encoder(target)
+        if weight is None:
+            weight = [1] * self.n_classes
+        assert inputs.size() == target.size(), 'predict {} & target {} shape do not match'.format(inputs.size(),
+                                                                                                  target.size())
+        loss = 0.0
+        for i in range(self.n_classes):
+            class_loss = self._metric_calculation(inputs[:, i], target[:, i])
+            loss += class_loss * weight[i]
+        return loss / self.n_classes
```

Comment:
I feel weird about this statement. I get that it should normalize the weights in the case that all the weights are 1. However, it feels wrong in cases where weights are explicitly given. As a user I would expect that my weights should add up to 1., but in that case they are scaled down again.  
Or am I missing sth. here?